### ISSUE TITLE
Support of custom Params or dynamic arrays

### ISSUE THREAD

**Issue by rongcuid** [rusqlite/rusqlite#1255]
I am currently stuck on dynamically generating `Params` for some custom types. This type has so many fields that the default implementations of `Params` do not contain a tuple large enough, and I have trouble making something that work.

I have a trait that has the following functions and types:

```rust
type Params: Params;
fn params(&self) -> Self::Params;
```

The only way to make it work is to use a tuple type for `Self::Params`, because none of the array types can satisfy the lifetime requirements (so &[&dyn ToSql]` does not work). `Box` or other managed memory types do not implement `Params` either (so no `Vec` or boxed array elements), and I cannot implement this trait for the type itself.

Is there a workaround?

**Comment by thomcc** [comment]
You should be able to use `rusqlite::params_from_iter`. See the documentation and example here: https://docs.rs/rusqlite/latest/rusqlite/struct.ParamsFromIter.html.

**Comment by rongcuid** [comment]
I don't think this is what I need. I know the number of params statically, but they have different types.

For instance, the params could be `(String, u64, Option<String>)` (of course, the real struct has a lot more).

The problem is that I can't make an iterator with different types, and boxed dyn types don't work with `Params`.

**Comment by rongcuid** [comment]
In fact, since there are only so many variable types to bind, can't rusqlite expose some sort of `BindCommand` enum for dynamic parameter binding?

**Comment by thomcc** [comment]
> boxed dyn types don't work with Params

This isn't true. We implement `ToSql` for `Box<dyn ToSql>` (the impl is https://docs.rs/rusqlite/latest/rusqlite/types/trait.ToSql.html#impl-ToSql-for-Box%3CT%3E).

**Comment by thomcc** [comment]
> can't rusqlite expose some sort of BindCommand enum for dynamic parameter binding?

Yes. I've considered this. But it's not needed for your case.

**Comment by thomcc** [comment]
For example <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=24868eca6bf3f85b322af0604722b01f>. This is just an example, and I'm not saying it's how you should structure things. But it does scale to arbitrary number of fields in your type.

**Comment by rongcuid** [comment]
Interesting, I must have missed something about `Box`. I will give it a try.

**Comment by rongcuid** [comment]
It is a lot of boilerplate, but I do get the types to check.

**Comment by rongcuid** [comment]
Well... I run into a problem where I get an error from the SQL statement (`out of range integral type conversion attempted` in this case), and I would wish the `Debug` trait is implemented for `Params`. I think it is possible.

**Comment by thomcc** [comment]
> and I would wish the Debug trait is implemented for Params. I think it is possible.

Params is a trait, so I'm not sure how this would work.

**Comment by rongcuid** [comment]
Since every "element" implements `ToSql`, it should be possible to convert each to int, string, etc. Those all have `Debug`.

**Comment by thomcc** [comment]
We still couldn't give a blanket impl of Debug for these types.

**Comment by gwenn** [comment]
> I am currently stuck on dynamically generating `Params` for some custom types. This type has so many fields that the default implementations of `Params` do not contain a tuple large enough, and I have trouble making something that work.

https://docs.rs/rusqlite/latest/rusqlite/trait.Params.html#positional-parameters
> Using the [rusqlite::params!](https://docs.rs/rusqlite/latest/rusqlite/macro.params.html) macro, e.g. thing.query(rusqlite::params![1, "foo", bar]). This is mostly useful for heterogeneous lists where the number of parameters greater than 16, or homogenous lists of parameters where the number of parameters exceeds 32.

Another alternative:
https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.raw_bind_parameter

**Comment by thomcc** [comment]
> [docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.raw_bind_parameter](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.raw_bind_parameter)

Oh I forgot we added this lol.

**Comment by rongcuid** [comment]
I have tried `params!` before. It creates a slice which cannot be returned due to lifetime issues.

As for `raw_bind_parameter`, this will not fit my case, because it is not the structs themselves that handle the statement/binding. Now, I can of course invert the control (and do something like `Trait::bind(stmt)`), and in that case I will just use `param![]`, but I am not sure if that is a good idea.

It would certainly be more convenient if I can just implement `Params` for a type.

**Comment by thomcc** [comment]
> It would certainly be more convenient if I can just implement Params for a type.

It's not out of the question, but it does need some design work. I'll think about it.

**Comment by rongcuid** [comment]
What I do right now is to "invert" the traits. Instead of returning a `Params` trait object and bind outside, I pass in the statement to a trait method for binding.

Before:

```
fn params(&self) -> Self::Params;
```

After:

```
fn bind(&self, stmt: &mut Statement);
```

Then I call `raw_bind_parameter` inside the method.
